### PR TITLE
Add new runtime call to crop a subregion from buffer_t

### PR DESCRIPTION
**Pull request by psuriana**
Modify ScheduleFunctions to call the _halide_buffer_crop runtime instead of creating an address_of of some call node.

### CODE REVIEW

**Comment by dsharletg** [comment]
This looks reasonable, but needs significant work to merge master.

File: src/ScheduleFunctions.cpp [hunk 1]
```diff
@@ -376,30 +376,33 @@ Stmt build_produce(Function f, const Target &target) {
...
-
-                BufferBuilder builder;
-                builder.host = host_ptr;
-                builder.type = f.output_types()[j];
-                builder.dimensions = f.dimensions();
-                int k = 0;
+                src_buf_name += ".buffer";
+                Expr src_buffer = Variable::make(type_of<struct halide_buffer_t *>(), src_buf_name);
+
+                Expr output_buffer_t = Call::make(type_of<struct halide_buffer_t *>(), Call::alloca, {(int)sizeof(halide_buffer_t)}, Call::Intrinsic);
```
Comment (abadams) [inline src/ScheduleFunctions.cpp]: I think it's worth wrapping these long lines.

File: src/ScheduleFunctions.cpp [hunk 2]
```diff
@@ -376,30 +376,33 @@ Stmt build_produce(Function f, const Target &target) {
...
-                    builder.extents.push_back(max - min + 1);
-                    builder.strides.push_back(stride);
+                    mins.push_back(min);
+                    extents.push_back(max - min + 1);
                 }
-                Expr output_buffer_t = builder.build();
+                args[3] = Call::make(Handle(), Call::make_struct, mins, Call::Intrinsic);
+                args[4] = Call::make(Handle(), Call::make_struct, extents, Call::Intrinsic);
+
+                output_buffer_t = Call::make(type_of<struct buffer_t *>(), Call::buffer_crop, args, Call::Extern);
```
Comment (abadams) [inline src/ScheduleFunctions.cpp]: should be struct halide_buffer_t *

File: src/runtime/buffer_t.cpp [hunk 3]
```diff
@@ -114,6 +114,22 @@ halide_buffer_t *_halide_buffer_init_from_buffer(halide_buffer_t *dst,
...
+                                     halide_dimension_t *dst_shape,
+                                     const halide_buffer_t *src,
+                                     const int *min, const int *extent) {
+    *dst = *src;
+    dst->dim = dst_shape;
+    for (int i = 0; i < dst->dimensions; i++) {
+        dst->dim[i] = src->dim[i];
+        dst->dim[i].min = min[i];
+        dst->dim[i].extent = extent[i];
+        dst->host += (min[i] - src->dim[i].min) * src->dim[i].stride * src->type.bytes();
```
Comment (abadams) [inline src/runtime/buffer_t.cpp]: This does a bunch of extra multiplies. Consider:

   int64_t offset = 0;
   for (int i ...) {
      offset += (min[i] - src->dim[i].min) * src->dim[i].stride;
   }
   dst->host += offset * src->type.bytes();

**Comment by abadams** [comment]
LGTM

**Pull request merged**